Lesson restructure (#24)
modular/max-llm-book: book/src/step_11.md

## Understanding weight loading

Weight loading involves three steps: loading the HuggingFace model, transferring weights to your MAX model, and transposing Conv1D weights.

First, load the pretrained model with `GPT2LMHeadModel.from_pretrained("gpt2")`. This downloads the weights (about 500MB) and returns a PyTorch model with the exact architecture you've implemented.

Next, transfer these weights to your MAX model using `max_model.load_state_dict(torch_model.state_dict())`. The `state_dict` is a dictionary mapping layer names to weight tensors. Since your MAX model has the exact same architecture and layer names, this transfer works seamlessly.

Finally, transpose the weights for layers that use Conv1D in HuggingFace: `c_attn`, `c_proj`, and `c_fc`. Conv1D stores weights in shape `[in_features, out_features]`, while Linear expects `[out_features, in_features]`. Use the `.T` property to transpose: `child.weight = child.weight.T`.

## Understanding model compilation

Before you can run text generation, compile the model with `.compile(token_type)`. Compilation analyzes the model's computation graph and generates optimized code for your hardware.

> You need to specify the input type using `TensorType`. This tells MAX what shape and dtype to expect:

Reviewer: ```suggestion
First, you need to specify the `token_type` input using `TensorType`. This tells the MAX compiler what shape and dtype to expect:
```

```python
token_type = TensorType(
    DType.int64,
    ("batch", "seqlen"),
    device=DeviceRef.from_device(device)
)
```

The shape uses symbolic dimensions `("batch", "seqlen")` rather than concrete numbers like `[1, 20]`. This allows the compiled model to handle any batch size and sequence length, not just fixed dimensions.

Compilation takes a few seconds but only happens once. After compilation, inference is much faster because MAX has optimized the entire computation graph.

## Understanding the tokenizer

The tokenizer converts text to token IDs and back. Load it with `GPT2Tokenizer.from_pretrained("gpt2")`, which downloads the same tokenization rules OpenAI used during training.

Set the padding token to match the end-of-sequence token: `tokenizer.pad_token = tokenizer.eos_token`. GPT-2 doesn't have a dedicated padding token, so we reuse the EOS token for this purpose.

The tokenizer provides two key methods:
- `tokenizer.encode(text)`: Converts text to a list of token IDs
- `tokenizer.decode(token_ids)`: Converts token IDs back to text

> Your helper functions `tokenize_text` and `decode_tokens` wrap these methods to convert between Python lists and MAX tensors.

Reviewer: Because we don’t use this code in this exercise (we already used them in step 9), I think it’s less likely to confuse if we just don’t mention them here.
```suggestion
Then pass the `tokenizer` to the `generate_text()` function you created
in step 10 (which passes it to `tokenize_text()` and `decode_tokens()`
from step 9).
```

## Implementing the main function

You'll implement the `main()` function that orchestrates the entire pipeline: loading models, transferring weights, initializing the tokenizer, compiling the model, and running an interactive prompt loop.

Start by loading the pretrained HuggingFace model:

```python

> torch_model = GPT2LMHeadModel.from_pretrained("gpt2")

Reviewer: Should this instead be called `hf_model`? I think that’s more clear if not also more accurate, because this is built from the transformers library (not strictly just PyTorch): https://huggingface.co/docs/transformers/en/model_doc/gpt2#transformers.GPT2LMHeadModel

</details>

> **Congratulations!** You've completed all 11 steps and built a complete GPT-2 implementation from scratch. Run `pixi run gpt2` to interact with your model through an interactive prompt and see your implementation in action!

Reviewer: I think this needs a bit more clarification about what this command does:

```suggestion
**Congratulations!** You've completed built a complete GPT-2 implementation from scratch. 

If code verification passed, you can execute your `step_11.py` code with `pixi run gpt2`. 
```
